plugins/node/opentelemetry-instrumentation-memcached/src/instrumentation.ts

```diff
@@ -0,0 +1,186 @@
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<typeof Memcached>(
+        'memcached',
+        ['>=2.2'],
+        (moduleExports, moduleVersion) => {
+          this._diag.debug(
+            `Patching ${Instrumentation.COMPONENT}@${moduleVersion}`
+          );
+          this.ensureWrapped(
+            moduleVersion,
+            moduleExports.prototype,
+            'command',
+            this.wrapCommand.bind(this, moduleVersion)
+          );
+          return moduleExports;
+        },
+        (moduleExports, moduleVersion) => {
+          this._diag.debug(
+            `Unpatching ${Instrumentation.COMPONENT}@${moduleVersion}`
+          );
+          if (moduleExports === undefined) return;
+          // `command` is documented API missing from the types
+          this._unwrap(moduleExports.prototype, 'command' as keyof Memcached);
+        }
+      ),
+    ];
+  }
+
+  wrapCommand(
+    moduleVersion: undefined | string,
+    original: (
+      queryCompiler: () => Memcached.CommandData,
+      server?: string
+    ) => any
+  ) {
+    const instrumentation = this;
+    return function (
+      this: Memcached,
+      queryCompiler: () => Memcached.CommandData,
+      server?: string
+    ) {
+      if (typeof queryCompiler !== 'function') {
+        return original.apply(this, arguments as any);
+      }
+      // The name will be overwritten later
+      const span = instrumentation.tracer.startSpan(
+        'unknown memcached command',
+        {
+          kind: api.SpanKind.CLIENT,
+          attributes: {
+            'memcached.version': moduleVersion,
+            ...Instrumentation.COMMON_ATTRIBUTES,
+          },
+        }
+      );
+      const context = api.trace.setSpan(api.context.active(), span);
+
+      return api.context.with(
+        context,
+        original,
+        this,
+        instrumentation.wrapQueryCompiler.call(
+          instrumentation,
+          queryCompiler,
+          this,
+          server,
+          context,
+          span
+        ),
+        server
+      );
+    };
+  }
+
+  wrapQueryCompiler(
+    original: () => Memcached.CommandData,
+    client: Memcached,
+    server: undefined | string,
+    context: api.Context,
+    span: api.Span
+  ) {
+    const instrumentation = this;
+    return function (this: Memcached) {
+      const query = original.apply(this, arguments as any);
+      const callback = query.callback;
+
+      span.updateName(`${query.type} ${query.key}`);
+      span.setAttributes({
+        'db.memcached.key': query.key,
+        'db.memcached.lifetime': query.lifetime,
+        [SemanticAttributes.DB_OPERATION]: query.type,
+        'db.statement': (instrumentation._config as InstrumentationConfig)
+          .enhancedDatabaseReporting
+          ? query.command
+          : undefined,
+        ...utils.getPeerAttributes(client, server, query),
+      });
+
+      query.callback = api.context.bind(
+        context,
```

Comment:
did not actually run it, but wouldn't setting the context this way make it so that every span started inside the callback will be the child of the (KIND=CLIENT) memcached span. 
Not sure if it's intentional, but I think this is something we try to avoid in instrumentations.

Reply:
Yeah, I was thinking about it as well while implementing it. In some cases it makes sense, in others, not.

What's the correct way to implement it, @vmarchaud , @obecny ?

Reply:
Since you use `.with` when calling `wrapQueryCompiler` you shouldn't need to .bind here (except if the memcached client is loosing context, for example if they have internal queues), if it works without i believe can you can remove

Reply:
@vmarchaud - do you expect the callback to have the `memcached` span context? or the parent context?

Reply:
parent context, we want to have the memcached context only within the internal code that does the request.

Reply:
I agree, but current implementation is setting the Memcached span as context for the callback.
So do we agree to change it? @Rauno56

Reply:
> Since you use `.with` when calling `wrapQueryCompiler` you shouldn't need to .bind here (except if the memcached client is loosing context, for example if they have internal queues), if it works without i believe can you can remove

@vmarchaud sorry I was reading your reply without full attention. I think it should still be binded, but to parent context, but not sure about it. better to try and see how it behaves